### PR TITLE
New solver: assert that normalizes-to hack RHS is always constrained after eval

### PR DESCRIPTION
**Pull request by compiler-errors**
See `NOTE` below, the `then_some` branch should never occur in practice.

r? @BoxyUwU

### CODE REVIEW

File: compiler/rustc_trait_selection/src/solve/eval_ctxt.rs [hunk 1]
```diff
@@ -285,18 +285,15 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
                             // The rhs of this `normalizes-to` must always be an unconstrained infer var as it is
                             // the hack used by `normalizes-to` to ensure that every `normalizes-to` behaves the same
                             // regardless of the rhs.
-                            //
-                            // However it is important not to unconditionally replace the rhs with a new infer var
-                            // as otherwise we may replace the original unconstrained infer var with a new infer var
-                            // and never propagate any constraints on the new var back to the original var.
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/eval_ctxt.rs:291]: I'm not sure if this ever matters in practice?

 Or to put it in another perspective, if this indeed *does* matter, I don't understand why it's then valid to replace a *constrained* RHS with a fresh var here without later equating that constrained RHS against that var.

File: compiler/rustc_trait_selection/src/solve/eval_ctxt.rs [hunk 2]
```diff
@@ -285,18 +285,15 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
...
-                            let term = this
-                                .term_is_fully_unconstrained(goal)
-                                .then_some(goal.predicate.term)
-                                .unwrap_or_else(|| {
-                                    this.next_term_infer_of_kind(goal.predicate.term)
-                                });
+
+                            // NOTE: This term should never be considered unconstrained, not least due to the fact
+                            // that the universe index has been bumped up at least once by `evaluate_goal` above.
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/eval_ctxt.rs:290]: why does `evaluate_goal` bump the universe index/alternatively is `term_is_fully_unconstrained` too pessimistic?

this feels worrying to me

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/solve/eval_ctxt.rs:290]: > I guess maybe we should change that code to replace the rhs with a new infer var, call evaluate_goal for the normalizes_to hack and then emit add an eq goal for the inference vars or sth? that feels safer to me

yeah i can do that, i have a branch that does that and coincidentally the code actually gets a bit cleaner too...

**Pull request closed without merging**